aspnetcore/fundamentals/websockets.md

```diff
@@ -148,13 +139,42 @@ When accepting the WebSocket connection before beginning the loop, the middlewar
 
 ::: moniker range=">= aspnetcore-2.2"
 
-### Handle client disconnects
+## WebSocket lifetime
+
+The WebSocket that is received from `AcceptWebSocketAsync` lives only as long as the middleware pipeline for the request is still running. When the middleware pipeline ends, the socket is closed. If your code continues to try to send and receive messages, a `WebSocketException` exception is raised with the following message:
+
+> The remote party closed the WebSocket connection without completing the close handshake.
+
+The exception has an `ObjectDisposedException` inner exception with the following message:
+
+> Cannot write to the response body, the response has completed. Object name: 'HttpResponseStream'.
+
+To avoid this error, don't let the request pipeline end before you're finished using the WebSocket. If you use a WebSocket from outside the request context (in a background thread or from another request) use code like the following example to make sure the request pipeline doesn't end prematurely:
+
+```csharp
+app.Use(async (context, next) => {
+    var socket = await context.WebSockets.AcceptWebSocketAsync();
+    var socketFinishedTcs = new TaskCompletionSource<object>();
+
+    BackgroundSocketProcessor.AddSocket(socket, socketFinishedTcs); 
+
+    await socketFinishedTcs.Task;
+});
+```
+
+The `await socketFinishedTcs.Task` statement keeps the the request open until the background processor calls `socketFinishedTcs.TrySetResult(null)` when it finishes with the socket.
+
+The same WebSocket closed error can happen if you return too soon from an MVC controller action method. If you accept a socket in the action method, make sure you wait for the code that uses the socket to complete before you return from the action method.
```

Comment:
Doesn't this apply to Razor methods too? If so, change to action method.

The same WebSocket closed error can happen if you return too soon from an ~MVC controller~ action method. If you accept a socket in the action method, make sure you wait for the code that uses the socket to complete before you return from the action method.